libc/string/string_cmp.c

```diff
@@ -0,0 +1,439 @@
+
+TEST_GROUP(string_memcmp);
+TEST_GROUP(string_strncmp);
+TEST_GROUP(string_strcmp);
+TEST_GROUP(string_strcoll);
+
+/*
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+*/
+
+
+TEST_SETUP(string_memcmp)
+{
+	/* Creating all needed data sets to run test cases */
+
+	nonZeroAsciiSet = testdata_createCharStr(BUFF_SIZE);
+	asciiSet = testdata_createCharStr(BUFF_SIZE);
+	hugeString = testdata_createCharStr(BUFF_SIZE * 4);
+	asciiSet[0] = 0;
+	memcpy(support, "Tests", sizeof("Tests"));
+}
+
+
+TEST_TEAR_DOWN(string_memcmp)
+{
+
+	free(nonZeroAsciiSet);
+	free(asciiSet);
+	free(hugeString);
+	memset(support, 0, sizeof(support));
+}
+
+
+TEST(string_memcmp, basic)
+{
+	TEST_ASSERT_EQUAL_INT(0, memcmp(basicStr, basicStr, sizeof(basicStr)));
+	TEST_ASSERT_LESS_THAN_INT(0, memcmp(basicStr, support, sizeof(basicStr)));
+	TEST_ASSERT_GREATER_THAN_INT(0, memcmp(support, basicStr, sizeof(basicStr)));
+}
+
+
+TEST(string_memcmp, unsigned_char_cast)
+{
+	/*
+	 * Checking do memcmp convert elements of input to unsigned char
+	 * If we get a value less than 0 that means memcmp treats the value of s1 as
+	 * signed char and return diff between SCHAR_MIN and 0
+	 * otherwise it will calculate the difference between 254 and 0 and memcmp will return
+	 * negative value
+	 */
+	charEdgeValue[0] = SCHAR_MIN;
+	TEST_ASSERT_GREATER_THAN_INT(0, memcmp(charEdgeValue, "\0", 1));
+	TEST_ASSERT_LESS_THAN_INT(0, memcmp("\0", charEdgeValue, 1));
+}
+
+
+TEST(string_memcmp, emptyInput)
+{
+	TEST_ASSERT_LESS_THAN_INT(0, memcmp(empty, asciiSet, BUFF_SIZE));
+	TEST_ASSERT_GREATER_THAN_INT(0, memcmp(asciiSet, empty, BUFF_SIZE));
+	TEST_ASSERT_LESS_THAN_INT(0, memcmp(empty, nonZeroAsciiSet, BUFF_SIZE));
+	TEST_ASSERT_GREATER_THAN_INT(0, memcmp(nonZeroAsciiSet, empty, BUFF_SIZE));
+	TEST_ASSERT_EQUAL_INT(0, memcmp(empty, empty, BUFF_SIZE));
+
+	/* Memory cmp is not sensitive for NUL characters */
+	TEST_ASSERT_NOT_EQUAL_INT(0, memcmp(empty, separated, sizeof(separated)));
+}
+
+
+TEST(string_memcmp, big)
+{
+	TEST_ASSERT_EQUAL_INT(0, memcmp(hugeString, hugeString, BUFF_SIZE * 4));
+	TEST_ASSERT_LESS_THAN_INT(0, memcmp(hugeZeroString, hugeString, BUFF_SIZE * 4));
+	TEST_ASSERT_GREATER_THAN_INT(0, memcmp(hugeString, hugeZeroString, BUFF_SIZE * 4));
+	TEST_ASSERT_EQUAL_INT(0, memcmp(hugeZeroString, hugeZeroString, BUFF_SIZE * 4));
+}
+
+
+TEST(string_memcmp, various_sizes)
+{
+	int i;
+
+	for (i = 1; i < BUFF_SIZE; i++) {
+		/* Checking various sizes of compare passed to memcmp */
+		TEST_ASSERT_LESS_THAN_INT(0, memcmp(asciiSet, nonZeroAsciiSet, i));
+
+		/*
+		 * The testing compares elements that value are every time on a different spot
+		 * and will check if the placement of the dataset on s1 and s2 will not impact on this
+		 */
+		support[i] = 1;
+		TEST_ASSERT_LESS_THAN_INT(0, memcmp(empty, support, BUFF_SIZE));
+		TEST_ASSERT_GREATER_THAN_INT(0, memcmp(support, empty, BUFF_SIZE));
+		support[i] = 0;
+	}
```

Comment:
Not tested, but this approach should be more effective.

```suggestion
	/* +1 to provide all ascii characters (ChrStr contains double 1 at the beginning) */
	const char *asciiStr = testdata_createCharStr(BUFF_SIZE + 1); //not sure whether const is possible there
	char asciiStr[BUFF_SIZE +1];
		
	memcpy(asciiStr2, asciiStr, sizeof(asciiStr));

	for (i = 1; i < BUFF_SIZE; i++) {
		asciiStr2[i] = asciiStr[i] - 1;
		TEST_ASSERT_EQUAL_INT(0, memcmp(asciiStr, asciiStr2, i));
		TEST_ASSERT_GREATER_THAN_INT(0, memcmp(asciiStr, asciiStr2, i + 1));
		asciiStr2[i] = asciiStr[i];
	}
	
	free(asciiStr);
	
```

 - you want to test especially cases, where different size changes the result of memcmp, that's why the approach above had been suggested.
 - comment not needed as it should be obvious that we check various sizes in such case,
 - there is no problem to use ascii string with NUL term, the exact data does not matter there, so I'd name those variables asciiStr
 - local variables seems to be better there, because we see the whole flow of data within them.